aeppl/loglik.py

```diff
@@ -87,30 +91,68 @@ def loglik(
 
     rng, size, dtype, *dist_params = rv_node.inputs
 
-    # Here, we plug the actual random variable into the log-likelihood graph,
-    # because we want a log-likelihood graph that only contains
-    # random variables.  This is important, because a random variable's
-    # parameters can contain random variables themselves.
-    # Ultimately, with a graph containing only random variables and
-    # "deterministics", we can simply replace all the random variables with
-    # their value variables and be done.
-
-    # tmp_rv_values = rv_values.copy()
-    # tmp_rv_values[rv_var] = rv_var
-
-    logpdf_var = _logpdf(rv_node.op, rv_value_var, *dist_params, **kwargs)
-
     # Replace random variables with their value variables
     replacements = rv_values.copy()
     replacements.update({rv_var: rv_value, rv_value_var: rv_value})
 
-    (logpdf_var,), _ = rvs_to_value_vars(
-        (logpdf_var,),
-        initial_replacements=replacements,
-    )
+    from collections import deque
+
+    # TODO: Walk the graph from its inputs to its outputs and construct the
+    # log
+    # TODO: Accept more than one graph.
+    # logp_fn = FunctionGraph(outputs=[rv_var], clone=False)
+
+    q = deque(io_toposort(graph_inputs((rv_var,)), (rv_var,)))
+
+    logpdf_var = None
+
+    while q:
+        node = q.popleft()
+
+        if not any(o in rv_values for o in node.outputs) and rv_var not in node.outputs:
+            if isinstance(node.op, RandomVariable):
+                warnings.warn(
+                    "Found a random variable that was neither among the observations "
+                    f"nor the conditioned variables: {node}"
+                )
+            continue
+
+        if isinstance(node.op, RandomVariable):
+            q_rv_var = node.outputs[1]
+            q_rv_value_var = replacements[q_rv_var]
+
+            # Replace `RandomVariable`s in the inputs with value variables.
+            # Also, store the results in the `replacements` map so that we
+            # don't need to redo these replacements.
+            value_var_inputs, _ = rvs_to_value_vars(
+                node.inputs,
+                initial_replacements=replacements,
+            )
+
+            q_logpdf_var = _logpdf(node.op, q_rv_value_var, *value_var_inputs, **kwargs)
+
+        else:
+            # Let's try deriving a log-likelihood from this
+            # non-`RandonVariable` node
+            q_logpdf_parts = []
+            for q_rv_var in node.outputs:
+                q_rv_value_var = replacements.get(q_rv_var, None)
 
-    if sum:
-        logpdf_var = at.sum(logpdf_var)
+                if q_rv_value_var is None:
+                    continue
+
+                # TODO: We need to prevent this from re-adding terms
```

Comment:
What is the TODO concern here? Do you have a MWE?

Reply:
That comment is referring to the cause of the failure in `test_loglik_subtensor`.  

When the `I = bernoulli(p)` is lifted through the `Y = normal(mu, sigma)` in the graph of `Y[I]`, producing `Y_new = normal(mu[I], sigma[I])`, the resulting log-likelihood ends up being something like `log(p(I) * (p(Y_new) * p(I)) * p(Y))`.  The term in parenthesis is introduced by the call to `_loglik`/`subtensor_loglik`, and the erroneous `p(I)` and `p(Y)` surrounding it are from the `io_toposort` walk through the original graph in `loglik`.

Since the `io_toposort` walk in `loglik` knows nothing about what `_loglik` has already done, it just keeps walking the nodes in `q` and "re-adds" the terms/nodes that were transformed and computed in `_loglik`/`subtensor_loglik`.